Add and document CRL APIs (#3523)
aws/s2n-tls: docs/USAGE-GUIDE.md

Additionally s2n-tls has functions for parsing certificate extensions on a certificate. Use `s2n_cert_get_x509_extension_value_length()` and `s2n_cert_get_x509_extension_value()` to obtain a specific DER encoded certificate extension from a certificate. `s2n_cert_get_utf8_string_from_extension_data_length()` and `s2n_cert_get_utf8_string_from_extension_data()` can be used to obtain a specific UTF8 string representation of a certificate extension instead. These functions will work for both RFC-defined certificate extensions and custom certificate extensions.

### Certificate Revocation

> Certificate revocation is a method in which CAs inform validators that an active certificate should not be trusted. This can occur when, for example, a private key has been leaked and the identity of the certificate's owner can no longer be trusted.

Author: Fixed!

> There are some circumstances in which a CA might wish to revoke one of its issued certificates. For instance, if an issued certificate’s private key is suspected to be leaked, the certificate should no longer be trusted. And thus, the CA will want to inform those validating certificate of the leak via a method of certificate revocation.

Author: I revised this section!

> s2n-tls supports two methods of certificate revocation validation: OCSP stapling, and CRLs. OCSP stapling differs from CRLs in that OCSP stapling is optional. A requester can ask a responder to send a stapled revocation status, but the responder may not actually send this response. However, CRLs exist for all valid certificates. If it can be assumed that OCSP stapling is supported by the peer of the application, it should be used as the method of certificate revocation validation. If this can not be assumed, and the CRLs for all received certificates can be obtained in advance, CRLs should be used instead.

Reviewer: ```suggestion
s2n-tls supports two methods of certificate revocation: OCSP stapling and CRLs. A fundamental difference between the two is that with OCSP stapling, the peer can choose not to send a certificate status response.  Applications using this method will have to decide whether or not to fail certificate validation in this case. In contrast, with CRLs the application can check the revocation status of any certificate encountered without relying on the peer. Users who want certificate revocation should look closely at their use-case and decide which method is more appropriate. We suggest using OCSP stapling if you're sure your peer supports OCSP stapling. Use CRLs if you're sure you can obtain CRLs for all certificates you expect to encounter.
```

I rewrote this to make the contrast between the two methods stronger.

Reviewer: Should we mention that OCSP stapling requires a lot less data than CRLs?

Author: This is much clearer. Thank you!

I added a sentence about CRLs needing more memory/network bandwidth. I also split this up into two paragraphs (let me know if you prefer it as one, it just seemed really long).

> s2n-tls supports two methods of certificate revocation validation: OCSP stapling, and CRLs. In general, OCSP stapling should be used when possible, and CRLs should be used in situations where OCSP stapling isn't supported or otherwise can't be used. For instance, if a server wants to determine if a client certificate is revoked, the server can request that the client send a stapled OCSP response, but there is no way to enforce that an unknown client will actually send the response. The client, for example, may not support OCSP stapling. As such, CRLs could be used in this case instead.

Author: I think this can most easily be summarized as the difference between OCSP being optional, and CRLs being always available to check. I revised this section with this distinction. Let me know if it's still unclear!

Additionally s2n-tls has functions for parsing certificate extensions on a certificate. Use `s2n_cert_get_x509_extension_value_length()` and `s2n_cert_get_x509_extension_value()` to obtain a specific DER encoded certificate extension from a certificate. `s2n_cert_get_utf8_string_from_extension_data_length()` and `s2n_cert_get_utf8_string_from_extension_data()` can be used to obtain a specific UTF8 string representation of a certificate extension instead. These functions will work for both RFC-defined certificate extensions and custom certificate extensions.

### Certificate Revocation

Certificate revocation is how CAs inform validators that an active certificate should not be trusted. This commonly occurs when a private key has been leaked and the identity of the certificate's owner can no longer be trusted.

> s2n-tls supports two methods of certificate revocation: OCSP stapling and CRLs. A fundamental difference between the two is that with OCSP stapling, the peer can choose not to send a certificate status response.  Applications using this method will have to decide whether or not to fail certificate validation in this case. In contrast, with CRLs the application can check the revocation status of any certificate encountered without relying on the peer. However, CRLs must be retrieved and stored by the application, which requires more network and memory utilization than OCSP stapling.

Reviewer: ```suggestion
s2n-tls supports two methods of certificate revocation: OCSP stapling and CRLs. A fundamental difference between the two is that with OCSP stapling, the peer can choose not to send a certificate status response. Applications using this method will have to decide whether or not to fail certificate validation in this case. In contrast, with CRLs the application can check the revocation status of any certificate encountered without relying on the peer. However, CRLs must be retrieved and stored by the application, which requires more network and memory utilization than OCSP stapling.
```
Nit: extra space

Author: Fixed, thanks!

The OCSP stapling information will be automatically validated if the underlying libcrypto supports OCSP validation. `s2n_config_set_check_stapled_ocsp_response()` can be called with "0" to turn this off. Call `s2n_connection_get_ocsp_response()` to retrieve the received OCSP stapling information for manual verification.

### CRL Validation

> Certificate Revocation Lists (CRLs) are lists of issued, unexpired certificates that have been revoked by the CA. CAs publish updated versions of these lists periodically. A peer wishing to verify a certificate can obtain a CRL from the CA, validate the CRL to make sure it was issued by the CA, and check to ensure the given certificate’s serial number is not contained in the list, and thus has not been revoked by the CA.

Reviewer 2: Because we now support both CRLs and OCSP, should we include a brief blurb on when you'd choose one over the other?

Maybe make a "Certificate Revocation" section and make both CRLs and OCSP into subsections.

Author: I added this!

> Certificate Revocation Lists (CRLs) are lists of issued, unexpired certificates that have been revoked by the CA. CAs publish updated versions of these lists periodically. A peer wishing to verify a certificate can obtain a CRL from the CA, validate the CRL to make sure it was issued by the CA, and check to ensure the given certificate’s serial number is not contained in the list, and thus has not been revoked by the CA.

Author: I believe this check is needed because all CRLs are obtained from CAs via HTTP, so you can't really be sure the CRL was retrieved from the CA or not. Also, there are cases where you indirectly obtain the CRL. Regardless, this is one of many CRL validation steps performed, so it may make sense for this to just say "obtain a CRL from the CA, validate the CRL, and check..."

> 

Author: I added a section about this to the new "certificate revocation" section which contains both CRL and OCSP now.

> To enable CRL validation in s2n-tls, the s2n CRL lookup callback must be implemented via `s2n_config_set_crl_lookup_callback()`. This callback is triggered once for each certificate received in the handshake. The received certificate's issuer hash can be retrieved in the callback via `s2n_crl_lookup_get_cert_issuer_hash()`. The issuer hash can be used to find the correct CRL for the given certificate.

Author: Added! Thanks!

> 

Reviewer 2: Is it worth discussing what the application needs to do? Like, what doesn't s2n-tls handle? What might "can be used to find the correct CRL" entail?

Author: I tried to add stuff here about what the application needs to do, in terms of getting the CRLs and implementing some way to look them up.

> The s2n CRL lookup callback expects a CRL to be returned via `s2n_crl_lookup_set()`. `s2n_crl`s can be created from PEMs via `s2n_crl_new()` and `s2n_crl_load_pem()`, and then freed via `s2n_crl_free()`. The issuer hash of the CRL can be obtained with `s2n_crl_get_issuer_hash()`. This issuer hash of a certificate will match the issuer hash of its CRL.

Reviewer: ```suggestion
Create a `s2n_crl` by calling `s2n_crl_new()` and add the obtained CRL data to the `s2n_crl` by calling `s2n_crl_load_pem()`.  Then call `s2n_crl_lookup_set()` to associate the CRL with a specific certificate in the callback. The `s2n_crl` should be freed via `s2n_crl_free()` when no longer needed. 

Call `s2n_crl_get_issuer_hash()` to get the issuer hash of the CRL if necessary. 
```

Not sure if I love my edit, but this is probably how I would write this paragraph. I switched the order of your instructions because they should be the order the user would call these functions(unless I'm mistaken).

I would also add this paragraph to the previous one as it is explaining what to do with the CRL data obtained.

Author: Switching the ordering definitely makes sense. I added some stuff about what the application should do here, and tried to make the ordering more correct.

> Calling `s2n_crl_lookup_set` will add the provided CRL to the list of CRLs that will be used to verify the certificate chain. Received certificates can also be ignored from the callback, via `s2n_crl_lookup_ignore`. When a certificate is ignored from the callback, no CRL is added to this list. If a certificate in the chain of trust cannot find a corresponding CRL in this list, the handshake will fail with a `S2N_ERR_CRL_LOOKUP_FAILED` error.

Reviewer: ```suggestion
Call `s2n_crl_lookup_ignore()` to ignore CRL data for a specific certificate in the callback. The certificate validation will fail with a `S2N_ERR_CRL_LOOKUP_FAILED` error if the ignored certificate is necessary for the chain of trust.
```

Author: This is definitely simpler. Thanks!

> Some received certificates may be extraneous. Extraneous certificates are additional certificates that aren't ultimately included in the chain of trust used to validate the end-entity certificate. If such a certificate is ignored from the callback, the handshake will proceed with no error.

Reviewer: You should add this paragraph to the previous one as this paragraph is still discussing s2n_crl_lookup_ignore.

Author: Fixed!

> By default, CRL timestamp validation is not performed. If a CRL returned from the CRL lookup callback is not yet active, or is expired, the CRL validation checks will continue and can succeed. If timestamp validation is desired, it should be performed in the CRL lookup callback with `s2n_crl_validate_active` and `s2n_crl_validate_not_expired`.

Reviewer: ```suggestion
By default, the CRL validation logic will not fail on CRLs that are not yet active, or expired CRLs. Perform this timestamp validation in the CRL lookup callback by calling `s2n_crl_validate_active()` and `s2n_crl_validate_not_expired()`.
```

Author: Fixed!

The OCSP stapling information will be automatically validated if the underlying libcrypto supports OCSP validation. `s2n_config_set_check_stapled_ocsp_response()` can be called with "0" to turn this off. Call `s2n_connection_get_ocsp_response()` to retrieve the received OCSP stapling information for manual verification.

#### CRL Validation

> Note: the CRL validation feature in s2n-tls is currently considered unstable, meaning the CRL APIs are subject to change in a future release. To access the CRL APIs, include `api/unstable/crl.h`.

> Certificate Revocation Lists (CRLs) are lists of issued, unexpired certificates that have been revoked by the CA. CAs publish updated versions of these lists periodically. A peer wishing to verify a certificate can obtain a CRL from the CA, validate the CRL, and check to ensure the given certificate’s serial number is not contained in the list, and thus has not been revoked by the CA.

Reviewer: ```suggestion
Certificate Revocation Lists (CRLs) are lists of issued, unexpired certificates that have been revoked by the CA. CAs publish updated versions of these lists periodically. A validator wishing to verify a certificate obtains a CRL from the CA, validates the CRL, and checks to ensure the certificate is not contained in the list, and therefore has not been revoked by the CA.
```

Author: Fixed. Thanks!

The s2n CRL lookup callback must be implemented and set via `s2n_config_set_crl_lookup_cb()` to enable CRL validation in s2n-tls. This callback will be triggered once for each certificate in the peer's certificate chain. 

> Applications must retrieve all CRLs for known certificates externally from s2n-tls. Applications should load the desired CRLs into memory, either in advance or in the callback, by creating `s2n_crl`s via `s2n_crl_new()`, and adding the obtained CRL data via `s2n_crl_load_pem()`. The `s2n_crl` should be freed via `s2n_crl_free()` when no longer needed.

Reviewer: Make more explicit what "known certificates" means.

Author: I tried to make this part more explicit. Let me know if it's still unclear!

> The CRLs for all certificates received in the handshake must be obtained in advance of the CRL lookup callback invocation, externally from s2n-tls. Applications should load these CRLs into memory, either in advance or in the callback, by creating `s2n_crl`s via `s2n_crl_new()`, and adding the obtained CRL data via `s2n_crl_load_pem()`. The `s2n_crl` should be freed via `s2n_crl_free()` when no longer needed.

Reviewer: ```suggestion
The CRLs for all certificates received in the handshake must be obtained in advance of the CRL lookup callback, outside of s2n-tls. Applications should load these CRLs into memory, by creating `s2n_crl`s via `s2n_crl_new()`, and adding the obtained CRL data via `s2n_crl_load_pem()`. The `s2n_crl` should be freed via `s2n_crl_free()` when no longer needed.
```
Just trimming some stuff down that I think is repetitive.

Author: Fixed!

> In the callback, call `s2n_crl_lookup_set()` to provide s2n-tls with a CRL to validate the certificate with. The application must implement a way to lookup the correct CRL for the given certificate. This lookup should use the certificate and CRL's issuer hashes, which will match. The certificate's issuer hash can be retrieved via `s2n_crl_lookup_get_cert_issuer_hash()`, and the CRL's issuer hash can be retrieved via `s2n_crl_get_issuer_hash()`.

Reviewer: These instructions should be in order of use.

Author: Agreed. I moved some of the APIs described here around to try and make this part more clear.

> The application must implement a way to look up the correct CRL for the given certificate. This lookup should use the hash of the received certificate's issuer, and the hash of the CRL's issuer, which will match. The certificate's issuer hash is retrieved via `s2n_crl_lookup_get_cert_issuer_hash()`, and the CRL's issuer hash is retrieved via `s2n_crl_get_issuer_hash()`. After the correct CRL is retrieved, call `s2n_crl_lookup_set()` to provide s2n-tls with this CRL, which will be used to validate the certificate with.

Reviewer: ```suggestion
The application must implement a way to look up the correct CRL for a given certificate. This can be done by comparing the hash of the received certificate's issuer with the hash of the CRL's issuer. The certificate's issuer hash is retrieved via `s2n_crl_lookup_get_cert_issuer_hash()`, and the CRL's issuer hash is retrieved via `s2n_crl_get_issuer_hash()`. Once a CRL is found with a matching issuer hash, call `s2n_crl_lookup_set()` to provide s2n-tls with this CRL.
```
More trimming and clarifying.

Author: Fixed. Thanks!

> Call `s2n_crl_lookup_ignore()` to ignore a specific certificate in the callback. Certificate validation will fail with a `S2N_ERR_CRL_LOOKUP_FAILED` error if the ignored certificate is necessary for the chain of trust. Some received certificates may be extraneous, in that they are additional certificates that aren't ultimately included in the chain of trust used to validate the end-entity certificate. If such a certificate is ignored from the callback, certificate validation will proceed with no error.

Reviewer: Change this paragraph to describe what to do if you can't find a CRL for the cert. Mention extraneous certs as well.

Author: I tried to make this more clear by describing the general case of what to do when a CRL can't be found, and then added sort of as an aside what happens when an ignored certificate is extraneous. Let me know it it's still unclear!

> If the CRL for a given certificate can't be found, call `s2n_crl_lookup_ignore()` to ignore this certificate. This will generally cause the certificate validation logic to fail with a `S2N_ERR_CRL_LOOKUP_FAILED` error. However, some received certificates are extraneous, and won't be included in the chain of trust. If an extraneous certificate is ignored, it will not cause the certificate validation logic to fail.

Reviewer: ```suggestion
Call `s2n_crl_lookup_ignore()` to ignore a received certificate if its CRL can't be found. This will cause the certificate validation logic to fail with a `S2N_ERR_CRL_LOOKUP_FAILED` error if the certificate is needed in the chain of trust. The certificate validation logic will not fail if the ignored certificate ends up not being included in the chain of trust.
```
Switching to more of an active voice.

Author: Thanks!!

The OCSP stapling information will be automatically validated if the underlying libcrypto supports OCSP validation. `s2n_config_set_check_stapled_ocsp_response()` can be called with "0" to turn this off. Call `s2n_connection_get_ocsp_response()` to retrieve the received OCSP stapling information for manual verification.

#### CRL Validation

> Note: the CRL validation feature in s2n-tls is currently considered unstable, meaning the CRL APIs are subject to change in a future release. To access the CRL APIs, include `api/unstable/crl.h`.

Certificate Revocation Lists (CRLs) are lists of issued, unexpired certificates that have been revoked by the CA. CAs publish updated versions of these lists periodically. A validator wishing to verify a certificate obtains a CRL from the CA, validates the CRL, and checks to ensure the certificate is not contained in the list, and therefore has not been revoked by the CA.

The s2n CRL lookup callback must be implemented and set via `s2n_config_set_crl_lookup_cb()` to enable CRL validation in s2n-tls. This callback will be triggered once for each certificate in the certificate chain. 

> The CRLs for all certificates received in the handshake must be obtained in advance of the CRL lookup callback, outside of s2n-tls. Applications should load these CRLs into memory, by creating `s2n_crl`s via `s2n_crl_new()`, and adding the obtained CRL data via `s2n_crl_load_pem()`. The `s2n_crl` should be freed via `s2n_crl_free()` when no longer needed.

Reviewer 2: Do the CRLs actually need to be obtained before the CRL callback? If you don't know what certs you're going to receive but don't mind the slowdown, couldn't you download the CRL during the callback?

Author: This is true in general, but not in s2n-tls currently. I added statements to a couple of sections to try and make this more clear.
